verilog/CST/type.cc

```diff
@@ -31,10 +33,33 @@ std::vector<verible::TreeSearchMatch> FindAllDataTypeDeclarations(
   return verible::SearchSyntaxTree(root, NodekDataType());
 }
 
+std::vector<verible::TreeSearchMatch> FindAllTypeDeclarations(
+    const verible::Symbol& root) {
+  return verible::SearchSyntaxTree(root, NodekTypeDeclaration());
+}
+
+std::vector<verible::TreeSearchMatch> FindAllStructDataTypeDeclarations(
+    const verible::Symbol& root) {
+  return verible::SearchSyntaxTree(root, NodekStructDataType());
+}
+
+std::vector<verible::TreeSearchMatch> FindAllUnionDataTypeDeclarations(
+    const verible::Symbol& root) {
+  return verible::SearchSyntaxTree(root, NodekUnionDataType());
+}
+
```

Comment:
Take a look at the cases of `kTypeDeclaration` in `verilog.y`:

https://github.com/google/verible/blob/d34420461f422d1cdbf4585a376b5a3210b9f828/verilog/parser/verilog.y#L3384

There are some cases where forward declarations are not covered, such as `typedef struct foo_t;`

If we choose to ignore those for the moment, make a TODO note, or file a follow-up issue.

If you wish to handle those now, consider this suggestion:
Make the forward declarations use the same form as their full declarations, like:

https://github.com/google/verible/blob/d34420461f422d1cdbf4585a376b5a3210b9f828/verilog/parser/verilog.y#L3461

but assign `nullptr` to the position where the brace-enclosed definition would normally go, for the sake of shape consistency.

and here:

https://github.com/google/verible/blob/d34420461f422d1cdbf4585a376b5a3210b9f828/verilog/parser/verilog.y#L3396

replace `$2` with a tagged node for `kStructDataType` that encloses `$2`.  (Same idea for unions)

I'm also amenable to other ideas as well, such as distinguishing forward-declarations from definitions with a new node enum.

Reply:
I see what you mean, I think it would be best to have forward declarations checked here as well.
I have changed the forward declarations to have the same structure as full declarations, but it looks quite fragile to me and I'm unsure if this is acceptable. I think the class and interface declarations would use a bit different structure anyway.